runtime/vm/jfr.cpp

```diff
@@ -0,0 +1,335 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "j9protos.h"
+#include "vm_internal.h"
+
+#if defined(J9VM_OPT_JFR)
+
+extern "C" {
+
+#undef DEBUG
+
+// TODO: allow different buffer sizes on different threads
+#define J9JFR_THREAD_BUFFER_SIZE (1024*1024)
+
+static void tearDownJFR(J9JavaVM *vm);
+static bool flushBufferToGlobal(J9VMThread *currentThread, J9VMThread *flushThread);
+static bool flushAllThreadBuffers(J9VMThread *currentThread);
+static U_8* reserveBuffer(J9VMThread *currentThread, UDATA size);
+static void jfrThreadCreated(J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
+static void jfrThreadDestroy(J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
+static void jfrClassesUnload(J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
+static void jfrVMShutdown(J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
+static void initializeEventFields(J9VMThread *currentThread, J9JFREvent *event, UDATA eventType, UDATA threadState);
+
+/**
+ * Flush a thread local buffer to the global buffer.
+ *
+ * The flushThread parameter must be either the current thread or be
+ * paused (e.g. by exclusive VM access).
+ *
+ * @param currentThread[in] the current J9VMThread
+ * @param flushThread[in] the J9VMThread to fluah
+ *
+ * @returns true on success, false on failure
+ */
+static bool
+flushBufferToGlobal(J9VMThread *currentThread, J9VMThread *flushThread)
+{
+#if defined(DEBUG)
+	PORT_ACCESS_FROM_VMC(currentThread);
+	j9tty_printf(PORTLIB, "\n!!! flushing %p of size %p\n", flushThread, flushThread->jfrData.bufferSize - flushThread->jfrData.bufferRemaining);
+#endif /* defined(DEBUG) */
+
+	// TODO: flush to global
```

Comment:
will this include synchronization to account for simultaneous `reserveBuffer` failures?

Reply:
Yes, access to the global buffer will need to be synchronized.